File: plugin/src/main/java/org/jukeboxmc/plugin/perms/manager/PermissionManagerImpl.java
package org.jukeboxmc.plugin.perms.manager;

import com.google.inject.Inject;
import com.google.inject.Singleton;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import lombok.RequiredArgsConstructor;
import org.jukeboxmc.plugin.perms.api.manager.PermissionManager;
import org.jukeboxmc.plugin.perms.repository.PermissionsRepository;

/**
 * @author Kaooot
 * @version 1.0
 */
@Singleton
@RequiredArgsConstructor(onConstructor_ = @Inject)
public class PermissionManagerImpl implements PermissionManager {

    private final PermissionsRepository repository;

    @Override
    public CompletableFuture<Boolean> addPermission(UUID uuid, String permission) {
        return this.addOrRemovePerm(uuid, permission, true);
    }

    @Override
    public CompletableFuture<Boolean> hasPermission(UUID uuid, String permission) {
        return this.repository.select(uuid)
            .thenApply(permissions -> permissions.contains(permission));
    }

    @Override
    public boolean hasCachedPermission(UUID uuid, String permission) {
        return this.repository.hasCachedPerm(uuid, permission);
    }

    @Override
    public CompletableFuture<Boolean> removePermission(UUID uuid, String permission) {
        return this.addOrRemovePerm(uuid, permission, false);
    }

    private CompletableFuture<Boolean> addOrRemovePerm(UUID uuid, String permission, boolean add) {
        return this.repository.select(uuid).thenApply(permissions -> {
            final boolean success;
            if (add) {
                success = permissions.add(permission);
            } else {
                success = permissions.remove(permission);
            }

            this.repository.update(uuid, permissions);

            return success;
        });
    }
}

File: plugin/src/main/kotlin/org/jukeboxmc/plugin/perms/manager/PermissionManagerImpl.kt
package org.jukeboxmc.plugin.perms.manager

import org.jukeboxmc.plugin.perms.api.manager.PermissionManager
import org.jukeboxmc.plugin.perms.repository.PermissionsRepository
import java.util.*
import java.util.concurrent.CompletableFuture

/**
 * @author Kaooot
 * @version 1.0
 */
class PermissionManagerImpl(private val repository: PermissionsRepository) : PermissionManager{

    override fun addPermission(uuid: UUID, permission: String): CompletableFuture<Boolean> {
        return this.addOrRemovePerm(uuid, permission, true)
    }

    override fun hasPermission(uuid: UUID, permission: String): CompletableFuture<Boolean> {
        return this.repository.select(uuid).thenApply { permissions: Set<String>? -> permissions?.contains(permission) ?: false }
    }

    override fun hasCachedPermission(uuid: UUID, permission: String): Boolean {
        return this.repository.hasCachedPerm(uuid, permission)
    }

    override fun removePermission(uuid: UUID, permission: String): CompletableFuture<Boolean> {
        return this.addOrRemovePerm(uuid, permission, false)
    }

    private fun addOrRemovePerm(uuid: UUID, permission: String, add: Boolean): CompletableFuture<Boolean> {
        return this.repository.select(uuid).thenApply { permissions: MutableSet<String>? ->
            val success: Boolean = if (add) {
                permissions?.add(permission) ?: false
            } else {
                permissions?.remove(permission) ?: false
            }

            if (permissions != null) {
                this.repository.update(uuid, permissions)
            }

            success
        }
    }
}